server/routes/basicDetails.ts

```diff
@@ -0,0 +1,411 @@
+    breachNotice.titleAndFullName = combinedName.trim().replace('  ', ' ')
+
+    // validation
+    const errorMessages: ErrorMessages = validateBasicDetails(breachNotice, req.body.dateOfLetter)
+    const hasErrors: boolean = Object.keys(errorMessages).length > 0
+
+    if (hasErrors) {
+      const alternateAddressOptions = initiateAlternateAddressSelectItemList(basicDetails, breachNotice)
+      const replyAddressOptions = initiateReplyAddressSelectItemList(basicDetails, breachNotice)
+      const basicDetailsDateOfLetter: string = req.body.dateOfLetter
+      res.render(`pages/basic-details`, {
+        errorMessages,
+        breachNotice,
+        basicDetails,
+        defaultOffenderAddress,
+        defaultReplyAddress,
+        alternateAddressOptions,
+        replyAddressOptions,
+        basicDetailsDateOfLetter,
+      })
+    } else {
+      // mark that a USER has saved the document at least once
+      breachNotice.basicDetailsSaved = true
+      await breachNoticeApiClient.updateBreachNotice(breachNoticeId, breachNotice)
+      console.log(req.body)
+      console.log(`action is ${req.body.action}`)
+
+      if (req.body.action === 'viewDraft') {
+        try {
+          await showDraftPdf(breachNotice.id, res)
+        } catch (err) {
+          // Render the page with the new error
+          errorMessages.pdfRenderError = {
+            text: 'There was an issue generating the draft report. Please try again or contact support.',
+          }
+
+          const alternateAddressOptions = initiateAlternateAddressSelectItemList(basicDetails, breachNotice)
+          const replyAddressOptions = initiateReplyAddressSelectItemList(basicDetails, breachNotice)
+          const basicDetailsDateOfLetter: string = req.body.dateOfLetter
+          res.render(`pages/basic-details`, {
+            errorMessages,
+            breachNotice,
+            basicDetails,
+            defaultOffenderAddress,
+            defaultReplyAddress,
+            alternateAddressOptions,
+            replyAddressOptions,
+            basicDetailsDateOfLetter,
+          })
+        }
+      } else if (req.body.action === 'saveProgressAndClose') {
+        res.setHeader('Content-Security-Policy', "script-src-elem 'unsafe-inline'")
+        res.send('<script>window.close();</script>')
+      } else {
+        res.redirect(`/warning-type/${req.params.id}`)
+      }
+    }
+  })
+
+  function validateBasicDetails(breachNotice: BreachNotice, userEnteredDateOfLetter: string): ErrorMessages {
+    const errorMessages: ErrorMessages = {}
+    const currentDateAtStartOfTheDay: LocalDateTime = LocalDate.now().atStartOfDay()
+    if (breachNotice.referenceNumber) {
+      if (breachNotice.referenceNumber.trim().length > 30) {
+        errorMessages.officeReferenceNumber = {
+          text: 'The Office Reference entered is greater than the 30 character limit.',
+        }
+      }
+    }
```

Comment:
Minor thing - you could simplify this if statement
```suggestion
    if (breachNotice.referenceNumber && breachNotice.referenceNumber.trim().length > 30) {
      errorMessages.officeReferenceNumber = {
        text: 'The Office Reference entered is greater than the 30 character limit.',
      }
    }
```